e2e/testcases/github_test.go

```diff
@@ -0,0 +1,285 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"testing"
+
+	"k8s.io/apimachinery/pkg/types"
+	"kpt.dev/configsync/e2e/nomostest"
+	"kpt.dev/configsync/e2e/nomostest/gitproviders"
+	"kpt.dev/configsync/e2e/nomostest/ntopts"
+	"kpt.dev/configsync/e2e/nomostest/policy"
+	nomostesting "kpt.dev/configsync/e2e/nomostest/testing"
+	"kpt.dev/configsync/pkg/api/configsync"
+	"kpt.dev/configsync/pkg/api/configsync/v1beta1"
+	"kpt.dev/configsync/pkg/core"
+	"kpt.dev/configsync/pkg/core/k8sobjects"
+	"kpt.dev/configsync/pkg/kinds"
+	"kpt.dev/configsync/pkg/status"
+)
+
+const (
+	githubAppRepoBranch              = "main"
+	githubAppRepoNamespace           = "github-app-test"
+	githubAppRepoConfigMap           = "test-cm"
+	githubAppRepoNamespacedConfigDir = "namespaced"
+)
+
+// Manual provisioning steps required for the githubapp tests to be runnable:
+// - Private Github repository created containing expected resources
+// - Github app created
+// - Github configured with read access to the private repository
+// - GitHub app configuration stored on local filesystem or GCP Secret Manager
+//
+// Expected format of repository (see above constants for expected NN values):
+//
+//	.
+//	├── LICENSE
+//	├── namespaced
+//	│         └── cm.yaml
+//	├── namespace-github-app-test.yaml
+//	└── README.md
+//
+// TODO: implement a gitprovider for github and replace these one-off tests
+func TestGithubAppRootSync(t *testing.T) {
+	nt := nomostest.New(t, nomostesting.SyncSource, ntopts.GitHubAppTest)
```

Comment:
This will require providing the flag with a separate prow job. Can we have a test case that creates a new RSync with the pre-provisioned repo and github app and validates the auth?

Then the test can be executed in all existing CI jobs.

Reply:
Yes, this is by design. Running this test requires specific resources to be manually provisioned, which in several general cases we shouldn't assume it has been:
- for local developer testing, we shouldn't assume an arbitrary developer has all of this manually provisioned. If a developer runs the hermetic kind tests with default options then they should succeed.
- for presubmits we shouldn't leak secrets, for the reasons I mentioned in the other comment. 

For these reasons, I think using a one-off test toggle like this is the most straightforward approach. It also makes transitioning to a `gitprovider` interface more straightforward.

Alternatively we could take on the following behavior, if we want to avoid creating a new ProwJob (still defaulting to `false`):
- If `--githubapp` is provided, run the whole test suite including the githubapp tests
- If `--githubapp` is not provided, run the whole test suite except for the githubapp tests

However, I don't see adding another ProwJob as a particularly major issue.

Reply:
ACK